backend/src/main/java/com/trackpack/app/service/ShipmentTrackingService.java

```diff
@@ -0,0 +1,137 @@
+        return repository.findAll();
+    }
+
+    public ShipmentTracking findById(UUID id) throws ResourceNotFoundException {
+        return repository.findById(id).
+                orElseThrow(() -> new ResourceNotFoundException("Parcel with id " + id + " is not found"));
+    }
+
+    public void add(ShipmentTracking shipmentTracking) {
+        repository.save(shipmentTracking);
+    }
+
+    public void delete(ShipmentTracking shipmentTracking) {
+        repository.delete(shipmentTracking);
+    }
+
+    public void deleteById(UUID id) {
+        repository.deleteById(id);
+    }
+
+
+    public void addCheckpoint(UUID id, CheckPoint checkPoint) throws ResourceNotFoundException {
+        ShipmentTracking shipmentTracking = findById(id);
+        List<CheckPoint> checkPoints = shipmentTracking.getCheckpoints();
+        if (checkPoints == null) {
+            checkPoints = new ArrayList<>();
+        }
+        checkPoints.add(checkPoint);
+    }
+
+    public void updateStatusInfo(UUID id, String status, OffsetDateTime statusChangeDate, String statusChangeReason)
+            throws ResourceNotFoundException {
+        ShipmentTracking shipmentTracking = findById(id);
+        if (status != null) {
+            shipmentTracking.setStatus(status);
+        }
+        shipmentTracking.setStatusChangeDate(statusChangeDate);
+        shipmentTracking.setStatusChangeReason(statusChangeReason);
+    }
+
+    public void updateEstimatedDeliveryDate(UUID id, OffsetDateTime estimatedDeliveryDate)
+            throws ResourceNotFoundException {
+        ZoneOffset zoneOffset = estimatedDeliveryDate.getOffset();
+        ZoneId zoneId = ZoneId.ofOffset("UTC", zoneOffset);
+        if (estimatedDeliveryDate.isAfter(OffsetDateTime.now(zoneId))) {
+            ShipmentTracking shipmentTracking = findById(id);
+            shipmentTracking.setEstimatedDeliveryDate(estimatedDeliveryDate);
+        } else {
+            throw new IllegalArgumentException("Delivery date must be greater than current date");
+        }
+    }
+
+    public void updateAddressFrom(UUID id, Address addressFrom) throws ResourceNotFoundException {
+        if(addressFrom == null){
+            throw new IllegalArgumentException("Address should not be a null value");
+        }
+            ShipmentTracking shipmentTracking = findById(id);
+            shipmentTracking.setAddressFrom(addressFrom);
+    }
+
+    public void updateAddressTo(UUID id, Address addressTo) throws ResourceNotFoundException {
+        if(addressTo == null){
+            throw new IllegalArgumentException("Address should not be a null value");
+        }
+            ShipmentTracking shipmentTracking = findById(id);
+            shipmentTracking.setAddressFrom(addressTo);
+    }
+
+    public void update(UUID id, Map<String, Object> changes) throws ResourceNotFoundException {
+        ShipmentTracking shipmentTracking = findById(id);
+        changes.forEach((field, value) -> {
```

Comment:
Why just not add ShipmentTracking parameter to method instead of Map and just check all fields for non-null values?